Implement newer RTU protocol of Growatt inverters

djwlindenaar opened the pull request:
# What does this implement/fix?

Newer Growatt inverters use a different set of registers. This commit adds the possibility to select RTU2 protocol as used by those newer inverters. At least the MIC series, MIN series and MAX series.
## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes esphome/issues#2997

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1976

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [X] ESP8266

## Example entry for `config.yaml`:
```yaml
# Example config.yaml
sensor:
  - platform: growatt_solar
    update_interval: 10s
    protocol_version: RTU2
    inverter_status:
      name: "Growatt Status Code"
    phase_a:
      voltage:
          name: "Growatt Voltage Phase"
      current:
          name: "Growatt Current Phase"
      active_power:
          name: "Growatt Power Phase"
    pv1:
      voltage:
          name: "Growatt PV1 Voltage"
      current:
          name: "Growatt PV1 Current"
      active_power:
          name: "Growatt PV1 Active Power"
    active_power:
      name: "Growatt Grid Active Power"
    pv_active_power:
      name: "Growatt PV Active Power"
    frequency:
      name: "Growatt Frequency"
    energy_production_day:
      name: "Growatt Today's Generation"
    total_energy_production:
      name: "Growatt Total Energy Production"
      accuracy_decimals: 1
    inverter_module_temp:
      name: "Growatt Inverter Module Temp"
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Commit:
Implement newer RTU protocol of Growatt inverters

Newer growatt inverters use a different set of registers. This commit adds the possibility to select RTU2 protocol as used by those newer inverters. At least the MIC series, MIN se
ries and MAX series.

Default behavior is unchanged, so no breaking change.

Select new protocol by adding to the growatt_solar section:
  protocol_version: RTU2

Commit:
Fix clang-format issues

Commit:
Fix lint issue introduced when fixing clang-format

jesserockz commented on esphome/components/growatt_solar/sensor.py:
```diff
@@ -152,6 +162,9 @@ async def to_code(config):
     await cg.register_component(var, config)
     await modbus.register_modbus_device(var, config)
 
+    if CONF_PROTOCOL_VERSION in config:
+        cg.add(var.set_protocol_version(config[CONF_PROTOCOL_VERSION]))
+
```
Having a default means it's always in the config
```suggestion
    cg.add(var.set_protocol_version(config[CONF_PROTOCOL_VERSION]))

```

jesserockz reviewed (changes requested):
Looks good, just one minor change

djwlindenaar committed:
Update esphome/components/growatt_solar/sensor.py

Since there is a default for CONF_PROTOCOL_VERSION, no need to check for existence in the config.

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

djwlindenaar commented:
> Looks good, just one minor change

Good point... Committed.

djwlindenaar committed:
Merge branch 'dev' into dev
